jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DelayedHandler.java

```diff
@@ -89,166 +147,200 @@ protected Callback getCallback()
-            if (!MimeTypes.Type.MULTIPART_FORM_DATA.is(contentTypeValue))
-                return null;
+        @Override
+        protected void delay()
+        {
+            CompletableFuture<Fields> futureFormFields = FormFields.from(getRequest(), _charset);
 
-            String boundary = MultiPart.extractBoundary(contentType);
-            if (boundary == null)
-                return null;
+            // if we are done already, then we are still in the scope of the original process call and can
+            // accept directly, otherwise we must execute a call to process as we are within a serialized
+            // demand callback.
+            futureFormFields.whenComplete(futureFormFields.isDone() ? this::accept : this::acceptAndExecute);
+        }
 
-            return new MultiPartDelayedProcess(next, request, response, callback, boundary);
+        public void accept(Fields fields, Throwable x)
+        {
+            if (x == null)
+                super.process();
+            else
+                Response.writeError(getRequest(), getResponse(), getCallback(), x);
         }
 
-        @Override
-        protected void delay(DelayedProcess request)
+        public void acceptAndExecute(Fields fields, Throwable x)
         {
-            request.run();
-            ((MultiPartDelayedProcess)request).whenDone();
+            if (x == null)
+                // We must execute here as even though we have consumed all the input, we are probably
+                // invoked in a demand runnable that is serialized with any write callbacks that might be done in process
+                getRequest().getContext().execute(super::process);
+            else
+                Response.writeError(getRequest(), getResponse(), getCallback(), x);
         }
+    }
 
-        protected static class MultiPartDelayedProcess extends DelayedProcess implements BiConsumer<MultiPartFormData.Parts, Throwable>
+    protected static class UntilMultiPartDelayedProcess extends DelayedProcess
+    {
+        private final MultiPartFormData _formData;
+
+        public UntilMultiPartDelayedProcess(Handler handler, Request wrapped, Response response, Callback callback, String contentType)
         {
-            private final MultiPartFormData _formData;
+            super(handler, wrapped, response, callback);
+            String boundary = MultiPart.extractBoundary(contentType);
+            _formData = boundary == null ? null : new MultiPartFormData(boundary);
+        }
 
-            public MultiPartDelayedProcess(Handler handler, Request wrapped, Response response, Callback callback, String boundary)
+        public void accept(MultiPartFormData.Parts parts, Throwable x)
+        {
+            if (x == null)
             {
-                super(handler, wrapped, response, callback);
-                _formData = new MultiPartFormData(boundary);
                 getRequest().setAttribute(MultiPartFormData.class.getName(), _formData);
```

Comment:
Move this to the constructor so it's done only once.

Reply:
I'm slightly worried about setting it before we have successfully parsed it.   But I can't see why, so OK.